dingo/gw/likelihood.py

```diff
@@ -247,22 +265,53 @@ def _log_likelihood(self, theta):
         Returns
         -------
         log_likelihood: float
+        dict containing detector matched filter and optimal SNR
         """
 
         # Step 1: Compute whitened GW strain mu(theta) for parameters theta.
+
+        # In the calibration-marginalization case, the values of mu for each detector have an additional leading
+        # dimension, which corresponds to the calibration draws. These are averaged over in computing the likelihood.
+
         mu = self.signal(theta)["waveform"]
         d = self.whitened_strains
 
-        # Step 2: Compute likelihood. log_Zn is precomputed, so we only need to
-        # compute the remaining terms rho2opt and kappa2
-        rho2opt = sum([inner_product(mu_ifo, mu_ifo) for mu_ifo in mu.values()])
-        kappa2 = sum(
-            [
-                inner_product(d_ifo, mu_ifo)
-                for d_ifo, mu_ifo in zip(d.values(), mu.values())
-            ],
-        )
-        return self.log_Zn + kappa2 - 1 / 2.0 * rho2opt
+        # Step 2: Compute likelihood. log_Zn is precomputed, so we only need to compute the remaining terms rho2opt
+        # and kappa2. This calculation batches multiple waveform draws.
+
+        rho2opt_detector = {
+            ifo: inner_product(mu_ifo, mu_ifo) for ifo, mu_ifo in mu.items()
+        }
+        kappa2_detector = {
+            ifo: inner_product(d[ifo], mu_ifo) for ifo, mu_ifo in mu.items()
+        }
+
+        # Calculate likelihood.
+
+        # First sum over detectors for each draw.
+        rho2opt = np.sum(list(rho2opt_detector.values()), axis=0)
+        kappa2 = np.sum(list(kappa2_detector.values()), axis=0)
+
+        if isinstance(rho2opt, np.ndarray):
+            # We marginalize the *likelihood*, not the *log likelihood*.
+            log_likelihoods = self.log_Zn + kappa2 - (1 / 2) * rho2opt
+            log_likelihood = logsumexp(log_likelihoods) - np.log(len(log_likelihoods))
+        else:
+            log_likelihood = self.log_Zn + kappa2 - (1 / 2) * rho2opt
+
+        # Calculate supplementary outputs, rho_mf and rho_opt, for each detector. By taking the mean, we marginalize
+        # over multiple draws, if present.
+
+        rho_opt_detector = {
+            f"{ifo}_optimal_snr": np.mean(v ** (1 / 2))
+            for ifo, v in rho2opt_detector.items()
+        }
+        rho_mf_detector = {
+            f"{ifo}_matched_filter_snr": np.mean(v / rho2opt_detector[ifo] ** (1 / 2))
+            for ifo, v in kappa2_detector.items()
+        }
+
+        return log_likelihood, rho_opt_detector | rho_mf_detector
```

Comment:
For added readability consider adding a line `rho_detectors_dict = rho_opt_detector | rho_mf_detector` and returning `rho_detectors_dict`.